DX-816: DDN workspace docs (#906)
hasura/ddn-docs: docs/private-ddn/ddn-workspace.mdx

---
sidebar_position: 5
sidebar_label: DDN workspace
description: "Learn how to develop with DDN workspace"
keywords:
  - hasura ddn
  - private ddn
  - dedicated vpc
  - enterprise ddn
  - create private ddn
  - ddn workspace
seoFrontMatterUpdated: false
---

import Thumbnail from "@site/src/components/Thumbnail";

# Developing with DDN workspace

## Introduction

> DDN workspace is a development environment to work with on Private DDN on the browser, without having to install DDN CLI and its dependencies in your local machine. It has a familiar VS Code like UI, runs on your data plane and has all the dependencies required to work with DDN pre-installed. 

Reviewer: ```suggestion
DDN workspace is a development environment for working with Private DDN in your browser, without having to install DDN CLI and its dependencies on your local machine. It has a familiar VS Code like UI, runs on your data plane and has all the dependencies required to work with DDN pre-installed. 
```

This guide will walk you through the steps for using DDN workspace to develop your API on a Data Plane in
[Private Hasura DDN](/private-ddn/overview.mdx). Before proceeding, you'll first need to be
[invited](/private-ddn/data-plane-collaboration.mdx) to a Data Plane.

:::info Prerequisite

> To provision workspaces in Private DDN, please reach out to us 

Reviewer: ```suggestion
To provision a new workspace in Private DDN, please reach out to us 
```

[here](https://hasura.io/contact-us).

:::

> The recommendation is to have a workspace dedicated per developer working on Private DDN API development. Once the workspace has been provisioned, it can be launched from the Private DDN page.

Reviewer: ```suggestion
The recommendation is to have a workspace dedicated per developer working on Private DDN API development. Once the workspace has been provisioned, it can be launched from the [Private DDN page](https://console.hasura.io/data-plane).
```

## Advantages

> - Get started with just your browser. No need to install DDN CLI and its dependencies (Docker) in your local machine.

Reviewer: ```suggestion
- Get started with just your browser. No need to install DDN CLI and its dependencies (e.g. Docker Engine) on your local machine.
```

- Network connectivity - As the workspace is running on your data plane, it has access to all the data sources and services available in the data plane network. No need to setup complex VPN/network policies to access your DBs from your local machine.
- Collaboration - Multiple developers can work on the same project simultaneously in the same workspace, if needed.

## Step 1. Launch a workspace

> To launch a workspace, click on the `Launch Workspace` button on the Private DDN page. This will open a new tab in your browser with the workspace UI. Click on the `Copy Password` button to copy the password to your clipboard. Use this password to login to the workspace.

Reviewer: ```suggestion
- To launch a workspace, click on the `Launch Workspace` button on the Private DDN. The list of all the Private DDN accessible to you are available [here](https://console.hasura.io/data-plane). This will open a new tab in your browser with the workspace UI.

- Click on the `Copy Password` button to copy the password to your clipboard. Use this password to login to the workspace.
```

Bring up the terminal using the shortcut `Ctrl` + `` ` ``. Run the following command to authenticate the CLI with a service account token. Follow [these](/project-configuration/project-management/service-accounts/#how-to-create-service-account) steps to create a service account token.

```bash
ddn auth login --access-token <service-account-token>
```
## Step 3. Scaffold out a new local project

```sh title="Next, create a new local project:"
ddn supergraph init my-project && cd my-project
```

Once you move into this directory, you'll see your project scaffolded out for you. You can view the structure by either
running `ls` in your terminal, or by opening the directory in the VS Code editor interface.

## Step 4. Add a data source

The command below will launch a wizard via the DDN CLI which will assist you in adding a data connector. You can learn
more about adding sources via data connectors [here](/data-sources/connect-to-a-source.mdx).

```sh title="Add a data source using the CLI:"
ddn connector init <connector_name> -i
```

> This will provision the necessary files — using your configuration — for connecting a data source to your supergraph

Reviewer: ```suggestion
This will provision the necessary files using your configuration for connecting a data source to your supergraph
```

API.

:::info Organize your data into subgraphs

You can also organize your API into **subgraphs**. Subgraphs are generally organized around team functions or business
outcomes and allow for independent ownership and governance of data sources. Learn more
[here](/project-configuration/subgraphs/index.mdx).

:::

## Step 5. Generate your metadata

```sh title="First, introspect your data source:"
ddn connector introspect <connector_name>
```

```sh title="Then, generate the metadata:"
ddn model add <connector_name> '*'
ddn command add <connector_name> '*'
ddn relationship add <connector_name> '*'
```

You'll see metadata files generated for each entity in your data source. You can learn more
[here](/data-modeling/overview.mdx).

:::info Note

> Local Development and launching the local console is not supported in DDN workspace. Once the metadata is generated, create a cloud project and deploy your supergraph to your Private DDN to query the API.

Reviewer: @SandeepSamba "local development" might have different meaning for different folks. And we are saying that DDN workspaces in dev environment. This looks a little bit weird.

How about saying `ddn supergraph build local` and ddn console --local is not supported in DDN workspace. Use ddn supergraph build create to create a project on cloud for building and testing your API.

Author: Done

:::

## Step 6. Create a project on your Data Plane

You'll create a new project using the DDN CLI and by passing two flags: the `data-plane-id` and the `plan`.

:::info Get Data Plane ID

You can find all the Data Planes you have access to on this page: https://console.hasura.io/data-plane

:::

:::info Available Plans

- ddn_free
- ddn_base
- ddn_advanced

Read more about plans [here](/reference/pricing.mdx)

:::

```sh title="Create a new project using your Data Plane ID and plan choice:"
ddn project init --data-plane-id <data-plane-id> --plan <plan-name>
```

The CLI will return information about your newly created project.

## Step 7. Build and deploy your supergraph

```sh title="build and deploy you supergraph:"
ddn supergraph build create
```

When this process is complete, the CLI will return a link to the hosted API where you can query your data.

## Step 8. Exporting your metadata

> You can export the metadata to your local machine by right clicking on the `Explorer` pane and selecting on the `Download` option , by. This will download the entire directory containing all the files in your workspace. Workspace also has `git` pre-installed, so you can configure it and push your changes to a remote repository as well. 

Reviewer: ```suggestion
You can export the metadata to your local machine by right clicking on the `Explorer` pane and selecting on the `Download` option. This will download the entire directory containing all the files in your workspace. Workspace also has `git` pre-installed, so you can configure it and push your changes to a remote repository as well. 
```

> You can export the metadata to your local machine by right clicking on the `Explorer` pane and selecting on the `Download` option , by. This will download the entire directory containing all the files in your workspace. Workspace also has `git` pre-installed, so you can configure it and push your changes to a remote repository as well. 

Reviewer: Instead of "in your workspace", this should be "on your local machine"?

Author: Clarified it

<Thumbnail src="/img/data-plane/export-metadata.png" alt="Export metadata" width="1000px" />

## Logout

> To logout from the workspace, choose on the `Sign out of code-server` option by clicking on the Menu button on the top left corner. This will close the workspace and you will be redirected to the Workspace login page. The state of your local project will be preserved and you can login again to continue working on it.

Reviewer: Hey - is it possible to say Sign Out of DDN Workspace (instead of using code-server) looks like some default value. Not sure if this help, but I found some config for vs code. [[here](https://github.com/coder/code-server/blob/34b8d2ed69811c3315a465f01492e9448c9254aa/ci/build/build-vscode.sh#L76)]
